client/lib/media/actions.js

```diff
@@ -84,84 +90,94 @@ MediaActions.fetchNextPage = function( siteId ) {
 	} );
 };
 
-MediaActions.add = function( siteId, file ) {
-	var query = {},
-		isUrl = 'string' === typeof file,
-		addHandler = isUrl ? 'addMediaUrls' : 'addMediaFiles',
-		fileUrl, transientMedia;
-
-	if ( Array.isArray( file ) || file instanceof window.FileList ) {
-		Array.prototype.slice.call( file ).forEach( MediaActions.add.bind( null, siteId ) );
-		return;
+MediaActions.add = function( siteId, files ) {
+	if ( files instanceof window.FileList ) {
+		files = [ ...files ];
 	}
 
-	// Generate a fake transient media item that can be rendered into the list
-	// immediately, even before the media has persisted to the server
-	transientMedia = {
-		ID: uniqueId( 'media-' ),
-		date: new Date().toISOString(),
-		transient: true
-	};
-
-	if ( 'string' === typeof file ) {
-		// Generate from string
-		assign( transientMedia, {
-			file: file,
-			extension: path.extname( file ).slice( 1 ),
-			mime_type: MediaUtils.getMimeType( file ),
-			title: path.basename( file )
-		} );
-	} else {
-		// Generate from window.File object
-		fileUrl = window.URL.createObjectURL( file );
-		assign( transientMedia, {
-			URL: fileUrl,
-			guid: fileUrl,
-			file: file.name,
-			extension: path.extname( file.name ).slice( 1 ),
-			mime_type: MediaUtils.getMimeType( file.name ),
-			title: path.basename( file.name ),
-			// Size is not an API media property, though can be useful for
-			// validation purposes if known
-			size: file.size
-		} );
+	if ( ! Array.isArray( files ) ) {
+		files = [ files ];
 	}
 
-	Dispatcher.handleViewAction( {
-		type: 'CREATE_MEDIA_ITEM',
-		siteId: siteId,
-		data: transientMedia
-	} );
-
-	// Abort upload if file fails to pass validation.
-	if ( MediaValidationStore.getErrors( siteId, transientMedia.ID ).length ) {
-		return;
-	}
-
-	// Assign parent ID if currently editing post
-	const post = PostEditStore.get();
-	if ( post && post.ID && ! isPlainObject( file ) ) {
-		file = {
-			parent_id: post.ID,
-			[ isUrl ? 'url' : 'file' ]: file
+	// We offset the current time when generating a fake date for the transient
+	// media so that the first uploaded media doesn't suddenly become newest in
+	// the set once it finishes uploading. This duration is pretty arbitrary,
+	// but one would hope that it would never take this long to upload an item.
+	const baseTime = Date.now() + ONE_YEAR_IN_MILLISECONDS;
```

Comment:
>That would be a very patient user to wait that long for an upload to complete.

Indeed! Maybe add an easter egg to give them a storage upgrade for such dedication :wink: